### PR TITLE
[doc] Contribute a new shape to declare custom palette tools

### PR DESCRIPTION
**Pull request by frouene**
# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [x] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: doc/iterations/2023.10/shapes/contribute_custom_frontend_tool.adoc [hunk 1]
```diff
@@ -0,0 +1,42 @@
+= (S) Contribute custom frontend tool
+
+== Problem
+
+Today there is no way to add custom complexe tool to a diagram palette.
```
Comment (pcdavid) [inline doc/iterations/2023.10/shapes/contribute_custom_frontend_tool.adoc]: Typo: complexe => complex.

File: doc/iterations/2023.10/shapes/contribute_custom_frontend_tool.adoc [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+* `canHandle(diagramId: string, diagramElementId: string)`->defined if the custom tool must be added to the palette.
+* `component(props: DiagramPaletteToolContributionComponentProps)`->the custom React component to add to the palette.
+
+With the following available props :
+
+[source]
+----
+editingContextId: string;
+diagramId: string;
+diagramElementId: string;
```
Comment (pcdavid) [inline doc/iterations/2023.10/shapes/contribute_custom_frontend_tool.adoc]: Can't we pass the actual `GQLDiagram` and `GQLDiagramElement` (which does not currently exist but should be a common supertype for `GQLNode` and `GQLEdge` which share a lot)?
I'm not sure the ids are enough to implement non-trivial tools.
For example, if I want a tool "Ungroup/unwrap" to be available only on diagram nodes which contain sub-nodes, can I _resolve_ the `diagramElementId` into the actual `GQLNode` to check for that?

**Review by pcdavid — DISMISSED** [review]

File: doc/iterations/2023.10/shapes/contribute_custom_frontend_tool.adoc (hunk 1 above)
Comment (frouene) [reply doc/iterations/2023.10/shapes/contribute_custom_frontend_tool.adoc]: Done

File: doc/iterations/2023.10/shapes/contribute_custom_frontend_tool.adoc (hunk 2 above)
Comment (frouene) [reply doc/iterations/2023.10/shapes/contribute_custom_frontend_tool.adoc]: Done, I added more available data to the canHandle like the kind for a node palette.

File: doc/iterations/2023.10/shapes/contribute_custom_frontend_tool.adoc [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+Add a new React component `DiagramPaletteToolContribution` with two props:
+
+* `canHandle(diagramId: string, diagramElementData: DiagramElementData)`->defined if the custom tool must be added to the palette.
+`DiagramElementData` is a union type, it'll depend on if it's a diagram palette or a node palette.
+** For a diagram palette, the programmer will have access to the diagram metadata:
+
+[source]
+----
+  id: string;
+  kind: string;
```
Comment (pcdavid) [inline doc/iterations/2023.10/shapes/contribute_custom_frontend_tool.adoc]: Won't this always be `siriusComponents://representation?type=Diagram`, and so not really useful?

Comment (frouene) [reply doc/iterations/2023.10/shapes/contribute_custom_frontend_tool.adoc]: It's part of the object `DiagramMetadata` that I reused for the props.

**Comment by frouene** [comment]
After discussion with @sbegaudeau, we only provided id, it's possible to retrieve data with the appropriate hook.

**Review by pcdavid — APPROVED** [review]

**Review by sbegaudeau — APPROVED** [review]

**Pull request merged**